express-notes-api/index.js

```diff
@@ -0,0 +1,86 @@
+const express = require('express');
+const app = express();
+const notes = require('./data.json');
+const fs = require('fs');
+app.listen(3000);
+app.get('/api/notes', (req, res) => {
+  const temp = [];
+  for (var k in notes.notes) {
+    temp.push(notes.notes[k]);
+  }
+  res.status(200);
+  res.json(temp);
+});
+app.get('/api/notes/:id', (req, res) => {
+  var id = Number(req.params.id);
+  if (id < 0 || Number.isNaN(id)) {
+    var errorMessage = { error: 'id must be a positive integer' };
+    res.status(400).send(errorMessage);
+  } else if (notes.notes[id] !== undefined) {
+    res.status(200).json(notes.notes[id]);
+  } else {
+    var notFoundMessage = { error: 'cannot find note with id ' + id };
+    res.status(404).send(notFoundMessage);
+  }
+});
+app.use(express.json());
+app.post('/api/notes', (req, res) => {
+  const newNotes = req.body;
+  if (newNotes.content !== undefined) {
+    newNotes.id = notes.nextId;
+    notes.nextId++;
+    notes.notes[newNotes.id] = newNotes;
+    var JSONNotes = JSON.stringify(notes, null, 2);
+    fs.writeFile('data.json', JSONNotes, err => {
+      console.error(err);
+      res.status(500).end();
+    });
+    res.status(201);
+    res.json(newNotes);
```

Comment:
Your success responses should be inside of your writeFile callback. writeFile is asynchronous, so we do not know when it finishes and if you respond like this with 201 or 200 or 204, it writeFile maybe failed. The code should be organized to only reply when writeFile finishes.

```js
fs.writeFile('data.json', JSONNotes, err => {
  if (err) {
    console.error(err);
    res.status(500).json({ error: 'an unexpected error occurred' });
  } else {
    res.status(201).json(newNote);
  }
})
```